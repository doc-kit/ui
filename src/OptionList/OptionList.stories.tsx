import { array, boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import * as React from 'react';
import { withReadme } from 'storybook-readme';
import OptionList from './OptionList';

// tslint:disable-next-line no-var-requires
const readme = require('./README.md');

storiesOf('OptionList', module)
  .addDecorator(withReadme(readme))
  .add('default', () => (
    <OptionList
      definitions={[
        {
          key: text('key', 'cancellation_reason'),
          required: boolean('required', false),
          type: text('type', 'string'),
          description: [
            'The reason given for the ',
            <a href="#" key="link">
              cancellation
            </a>,
            ' by the member support employee.',
          ],
          default: text('default value', 'null'),
          values: array('possible values', [
            'duplicate',
            'fraudulent',
            'requested_by_customer',
          ]),
        },
      ]}
    />
  ))
  .add('multiple options', () => (
    <OptionList
      definitions={[
        {
          key: 'product_returned',
          required: false,
          type: 'boolean',
          description:
            'Whether or not the product has been returned by the customer.',
          default: 'false',
        },
        {
          key: 'status',
          required: true,
          type: 'string',
          description: [
            'Determines the current ',
            <a href="#" key="link">
              activity status
            </a>,
            ' of the account holder.',
          ],
          default: 'active',
          values: ['active', 'inactive'],
        },
      ]}
    />
  ));
