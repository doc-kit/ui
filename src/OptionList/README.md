# Option List

Useful to render options, parameters, or arguments in a stylish list structure.

## Features

- Automatically builds the sentence format for the option based on the props passed into the component.
- Accepts pattern to allow HTML such as links to be built into the description sentence.

## Usage

**Import**:

`import { OptionList } from '@doc-kits/react';`

**JSX**:

```js
<OptionList
  tableTitle="Arguments",
  definitions={[
    {
      key: "cancellation_reason",
      required: true,
      type: "string",
      description: "The reason given for the cancellation."
      default: "null",
      values: [
        "duplicate",
        "fraudulent",
        "requested_by_customer",
      ],
    },
    {...},
  ]}
/>
```

## Props

| Property               | Type                                | Required | Default | Description                              |
| ---------------------- | ----------------------------------- | -------- | ------- | ---------------------------------------- |
| tableTitle             | string                              | no       | Options | Sets the title for the table             |
| definitions            | Array[Objects]                      | yes      | null    | A list of option properties              |
| definition.key         | string                              | yes      | null    | A name for the key of the option         |
| definition.required    | boolean                             | no       | false   | Set true if this option is required      |
| definition.type        | string                              | yes      | null    | The type for the option                  |
| definition.description | string, Array[string / HTMLElement] | no       | null    | A description for the option             |
| definition.default     | string                              | no       | null    | The default value of the option          |
| definition.values      | Array[string]                       | no       | null    | A list of possible values for the option |

## Styles

The default styles injected into `OptionList` are shown below.

```js
const styles = {
  wrapper: {
    borderBottom: '1px solid #cbd8de',
    color: '#4a6271',
    fontSize: ['0.875rem', '1rem'],
    lineHeight: '1.5',
  },
  title: {
    color: '#4a6271',
    fontSize: 'inherit',
    borderBottom: '1px solid #cbd8de',
    paddingBottom: '0.313em',
    textTransform: 'uppercase',
  },
  option: {
    borderTop: '1px solid #edf3f3',
    padding: '1.143em 0',
  },
  key: {
    color: '#334049',
    fontSize: 'inherit',
    fontWeight: '500',
  },
  optional: {
    color: '#939da3',
    fontSize: '1em',
  },
  required: {
    color: '#ffae54',
    fontSize: '0.857em',
    border: '1px solid #ffae54',
    borderRadius: '10px',
    padding: '0 0.571em',
    textTransform: 'uppercase',
  },
  type: {
    color: '#939da3',
    fontSize: 'inherit',
  },
  description: {
    color: 'inherit',
    fontSize: 'inherit',
    margin: '1em 0 0 0',
  },
  values: {
    color: '#939da3',
  },
  code: {
    backgroundColor: '#fafcfc',
    borderRadius: '3px',
    border: `1px solid ${darken(0.05, '#fafcfc')}`,
    color: '#b93d6a',
    padding: '0 0.4em',
    fontFamily: 'inherit',
  },
  palette: {
    link: '#0099e5',
  },
};
```

## Notes

- You will _almost always_ want to add a punctuation period at the end of your `description` as it will be entered as a sentence.

- You may want to add a link into the description. To do so, you must use a mixed Array of strings and an `a` HTMLElement. It would look like the following: <br><br>

  ```js
  {
    description: [
      'The reason given for the ',
      <a href="#" key="link">cancellation</a>,
      ' by the member support employee.',
    ],
  }
  ```
