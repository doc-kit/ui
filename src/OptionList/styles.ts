import { darken } from 'polished';
import { base, palette, ui } from '../toolkit/theme';

const codeBg = '#fafcfc';

const styles = {
  wrapper: {
    ...base,
    borderBottom: `1px solid ${palette.grayScale['200']}`,
    color: ui.text.color,
    fontSize: ['0.875rem', '0.875rem', '1rem'],
    lineHeight: '1.5',
  },
  title: {
    borderBottom: `1px solid ${palette.grayScale['200']}`,
    color: ui.text.color,
    fontSize: 'inherit',
    paddingBottom: '0.313em',
    textTransform: 'uppercase',
  },
  option: {
    borderTop: `1px solid ${palette.grayScale['100']}`,
    padding: '1.143em 0',
  },
  key: {
    color: '#334049',
    fontSize: 'inherit',
    fontWeight: '600',
  },
  optional: {
    color: palette.grayScale['400'],
  },
  required: {
    border: `1px solid ${palette.primary.orange}`,
    borderRadius: '10px',
    color: palette.primary.orange,
    fontSize: '0.875em',
    padding: '0 0.571em',
    textTransform: 'uppercase',
  },
  type: {
    color: palette.grayScale['400'],
    fontSize: 'inherit',
  },
  description: {
    color: 'inherit',
    fontSize: 'inherit',
    margin: '1em 0 0 0',
  },
  values: {
    color: palette.grayScale['400'],
  },
  code: {
    backgroundColor: codeBg,
    border: `1px solid ${darken(0.05, codeBg)}`,
    borderRadius: '3px',
    color: palette.primary.red,
    fontFamily: 'inherit',
    padding: '0 0.4em',
  },
};

export default styles;
