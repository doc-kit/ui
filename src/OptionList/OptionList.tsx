/** @jsx jsx */
import { jsx } from '@emotion/core';
import styled from '@emotion/styled';
import * as merge from 'deepmerge';
import { ThemeProvider } from 'emotion-theming';
import { mq } from '../toolkit/responsive';
import settings from '../toolkit/settings';
import styles from './styles';

interface Definition {
  key: string;
  required?: boolean;
  type: string;
  description?: string | Array<string | any>;
  default?: string;
  values?: string[];
}

interface Props {
  tableTitle?: string;
  definitions: Definition[];
  localStyles?: any;
}

interface RequiredProps {
  required?: boolean;
}

const Wrapper = styled.div`
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;

  ${p => p.theme.mq(p.theme.wrapper)};

  a {
    color: ${p => p.theme.colors.link};
    text-decoration: none;

    &:hover {
      text-decoration: underline;
    }

    &:active,
    &:visited {
      color: ${p => p.theme.colors.link};
    }
  }
`;

const Title = styled.div`
  ${p => p.theme.mq(p.theme.title)};
`;

const Option = styled.div`
  ${p => p.theme.mq(p.theme.option)};

  &:first-of-type {
    border-top: none;
  }
`;

const Info = styled.div`
  align-items: baseline;
  display: flex;

  div {
    margin-right: 0.357em;

    &:last-of-type {
      margin-right: 0;
    }
  }
`;

const Key = styled.div`
  ${p => p.theme.mq(p.theme.key)};
`;

const Required = styled.div<RequiredProps>`
  border: ${p => (p.required ? p.theme.required.border : 'none')};
  border-radius: ${p => p.theme.required.borderRadius};
  color: ${p => (p.required ? p.theme.required.color : p.theme.optional.color)};
  font-size: ${p =>
    p.required ? p.theme.required.fontSize : p.theme.optional.fontSize};
  padding: ${p => (p.required ? p.theme.required.padding : '0')};
  text-transform: ${p =>
    p.required ? p.theme.required.textTransform : 'none'};
`;

const Type = styled.div`
  ${p => p.theme.mq(p.theme.type)};
`;

const Description = styled.p`
  ${p => p.theme.mq(p.theme.description)};
`;

const Values = styled.span`
  ${p => p.theme.mq(p.theme.values)};

  span:not(:first-of-type)::before {
    content: ', ';
  }

  span:nth-last-of-type(2):not(:first-of-type)::after {
    content: ',';
  }

  span:last-child:not(:only-of-type)::before {
    content: ' or ';
  }

  div:nth-of-type(1) {
    margin-bottom: 0.25em;
  }
`;

const Code = styled('code')`
  ${p => p.theme.mq(p.theme.code)};
`;

const OptionList = ({ tableTitle, definitions, localStyles }: Props) => {
  const sortedDefinitions = definitions.sort(
    (a: any, b: any) => b.required - a.required
  );
  const componentStyles = (appTheme: object = {}) =>
    merge.all([settings, styles, { mq }, appTheme, localStyles]);

  return (
    <ThemeProvider theme={componentStyles}>
      <Wrapper>
        <Title>{tableTitle}</Title>

        {sortedDefinitions.map(argument => {
          const requiredText = argument.required ? 'required' : 'optional';

          return (
            <Option key={argument.key}>
              <Info>
                <Key>{argument.key}</Key>
                <Required required={argument.required}>{requiredText}</Required>
                <Type>{argument.type}</Type>
              </Info>
              <Description>
                <span>{argument.description} </span>
                {argument.default && (
                  <span>
                    The default value is <Code>{argument.default}</Code>.
                  </span>
                )}
                {argument.values && <span> Possible values are </span>}
                {argument.values && (
                  <Values>
                    {argument.values.map((value: string, index: number) => (
                      <span key={index.toString()}>
                        <Code>{value}</Code>
                      </span>
                    ))}
                    .
                  </Values>
                )}
              </Description>
            </Option>
          );
        })}
      </Wrapper>
    </ThemeProvider>
  );
};

OptionList.defaultProps = {
  tableTitle: 'Options',
  localStyles: {},
  definitions: {
    required: false,
  },
};

export default OptionList;
