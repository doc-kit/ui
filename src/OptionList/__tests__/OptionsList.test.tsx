import React from 'react';
import renderer from 'react-test-renderer';
import OptionList from '../OptionList';

jest.mock('../../toolkit/constructTheme', () => {
  return () => ({
    colors: {},
    required: {},
    optional: {},
    mq: () => undefined,
  });
});

describe('<OptionList />', () => {
  it('should render with minimum requirements', () => {
    const tree = renderer
      .create(
        <OptionList
          definitions={[
            {
              key: 'align-content',
              type: 'string',
            },
          ]}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render with all available passed in options', () => {
    const tree = renderer
      .create(
        <OptionList
          tableTitle="Arguments"
          styles={{ title: { color: 'green' } }}
          definitions={[
            {
              key: 'align-content',
              type: 'string',
              required: true,
              description: [
                'The reason given for the ',
                <a href="#" key="link">
                  cancellation
                </a>,
                ' by the member support employee.',
              ],
              default: 'stretch',
              values: ['stretch', 'start', 'end'],
            },
          ]}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
