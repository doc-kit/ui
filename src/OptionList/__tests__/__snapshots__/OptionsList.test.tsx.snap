// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OptionList /> should render with all available passed in props 1`] = `
<div
  className="css-n5bgog"
>
  <div
    className="css-n34rpg"
  >
    <div
      className="css-1h6g3aq"
    >
      Arguments
    </div>
    <div
      className="css-1fcz6ig"
    >
      <div
        className="css-jxhuyt"
      >
        <div
          className="css-p5jona"
        >
          align-content
        </div>
        <div
          className="css-xurkgk"
        >
          required
        </div>
        <div
          className="css-14spoyu"
        >
          string
        </div>
      </div>
      <p
        className="css-1gks27d"
      >
        <span>
          The reason given for the 
          <a
            href="#"
          >
            cancellation
          </a>
           by the member support employee.
           
        </span>
        <span>
          The default value is
           
          <code
            className="css-eypjzh"
          >
            stretch
          </code>
          .
        </span>
        <span>
           Possible values are 
        </span>
        <span
          className="css-p5d22a"
        >
          <span>
            <code
              className="css-eypjzh"
            >
              stretch
            </code>
          </span>
          <span>
            <code
              className="css-eypjzh"
            >
              start
            </code>
          </span>
          <span>
            <code
              className="css-eypjzh"
            >
              end
            </code>
          </span>
          .
        </span>
      </p>
    </div>
  </div>
</div>
`;

exports[`<OptionList /> should render with minimum requirements 1`] = `
<div
  className="css-n5bgog"
>
  <div
    className="css-n34rpg"
  >
    <div
      className="css-1sb2enk"
    >
      Options
    </div>
    <div
      className="css-1fcz6ig"
    >
      <div
        className="css-jxhuyt"
      >
        <div
          className="css-p5jona"
        >
          align-content
        </div>
        <div
          className="css-uua7zr"
        >
          optional
        </div>
        <div
          className="css-14spoyu"
        >
          string
        </div>
      </div>
      <p
        className="css-1gks27d"
      >
        <span>
           
        </span>
      </p>
    </div>
  </div>
</div>
`;
